### PR TITLE
feat: Implement Interrupt Descriptor Table and Interrupt Routines 

### PR DESCRIPTION
**Pull request by raphfaur**
This PR aims to implement utilities to create interrupt routines, load them in an IDT, and load this IDT

### CODE REVIEW

File: src/bios/f-init32/Cargo.toml [hunk 1]
```diff
@@ -6,7 +6,9 @@ edition = "2021"
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
-flib = { path = "../flib/", features = ["alloc"] }
+flib = { path = "../flib/", features = ["alloc"]}
+f-macros = {path = "../f-macros/"}
```
Comment (sleiderr) [inline src/bios/f-init32/Cargo.toml:10]:
```suggestion
flib = { path = "../flib/", features = ["alloc"] }
f-macros = { path = "../f-macros/" }
```

File: src/bios/f-init32/Cargo.toml [hunk 2]
```diff
@@ -6,7 +6,9 @@ edition = "2021"
...
+numtoa = "0.2.4"
```
Comment (sleiderr) [inline src/bios/f-init32/Cargo.toml:11]:
```suggestion
numtoa = "0.2"
```

File: src/bios/f-init32/src/interrupts/handlers.rs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    print_str("Lol", 0);
+    loop {
+
+    }
+
+    // scheduler_ref!(scheduler);
+    // let action = ScheduledAction::new(0x00);
+    // scheduler.schedule(action);
+}
+
```
Comment (sleiderr) [inline src/bios/f-init32/src/interrupts/handlers.rs:72]: faut retirer ça non ?
```suggestion
```

File: src/bios/f-macros/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+name = "f-macros"
+version = "0.1.0"
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+syn = { version = "2.0.28", features = ["full"]}
+quote = "1.0.32"
+proc-macro2 = "1.0.66"
```
Comment (sleiderr) [inline src/bios/f-macros/Cargo.toml:11]:
```suggestion
proc-macro2 = "1.0"
```

File: src/bios/f-macros/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,13 @@
+[package]
+name = "f-macros"
+version = "0.1.0"
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+syn = { version = "2.0.28", features = ["full"]}
+quote = "1.0.32"
```
Comment (sleiderr) [inline src/bios/f-macros/Cargo.toml:10]:
```suggestion
quote = "1.0"
```

File: src/bios/f-macros/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,13 @@
+[package]
+name = "f-macros"
+version = "0.1.0"
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+syn = { version = "2.0.28", features = ["full"]}
```
Comment (sleiderr) [inline src/bios/f-macros/Cargo.toml:9]:
```suggestion
syn = { version = "2.0.28", features = ["full"] }
```

File: src/bios/f-macros/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+            table.write(#offset);
+        }
+    };
+
+    stream.into()
+}
+
+/// This proc macro aims to provide a higher level interface for interrupts definition.
+/// It will soon support gate type parameter in order to adapt asm! instructions for
+/// specific cases
```
Comment (sleiderr) [inline src/bios/f-macros/src/lib.rs:106]:
```suggestion
/// specific cases.
```

File: src/bios/flib/src/idt/idt.rs [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+/// | --------------------- |
+/// |  Offset + 0 | entry 0 |
+/// |  Offset + 8 | entry 1 |
+///  -----------------------
+pub struct Table {
+    entries: Vec<GateDescriptor>,
+}
+
+impl Table {
+    /// Writes the [Table] to a given offset. You have to ensure that there is enough
```
Comment (sleiderr) [inline src/bios/flib/src/idt/idt.rs:39]:
```suggestion
    /// Writes the `Table` to a given offset. You have to ensure that there is enough
```
Je suis d'avis d'éviter les liens vers elle-même quand on commente des élements de la struct

File: src/bios/flib/src/idt/idt.rs [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+    /// Creates an empty [Table]
+    pub fn empty() -> Self {
+        Self {
+            entries: Vec::new(),
+        }
+    }
+
+    /// Returns a mutable reference to a [GateDescriptor] wrapped in an [Option].
+    /// Index starts at 0.
+    pub fn get_entry_mut(&mut self, index : usize) -> Option<&mut GateDescriptor>{
```
Comment (sleiderr) [inline src/bios/flib/src/idt/idt.rs:59]:
```suggestion
    pub fn get_entry_mut(&mut self, index : usize) -> Option<&mut GateDescriptor> {
```

File: src/bios/flib/src/idt/idt.rs [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+        self.entries.get_mut(index)
+    }
+
+    /// Returns an immutable reference to a [GateDescriptor] wrapped in an [Option].
+    /// Index starts at 0.
+    pub fn get_entry(&self, index : usize) -> Option<&GateDescriptor>{
```
Comment (sleiderr) [inline src/bios/flib/src/idt/idt.rs:65]:
```suggestion
    pub fn get_entry(&self, index : usize) -> Option<&GateDescriptor> {
```

File: src/bios/flib/src/idt/idt.rs [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+        self.entries.get(index)
+    }
+
+    /// Populates the [Table] with default [`GateDescriptor`] to reach 256 entries (required)
```
Comment (sleiderr) [inline src/bios/flib/src/idt/idt.rs:69]:
```suggestion
    /// Populates the `Table` with default [`GateDescriptor`] to reach 256 entries (required)
```

File: src/bios/flib/src/idt/idt.rs [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+    pub fn populate_default(&mut self) {
+        let mut i = self.len();
+        while i < 256 {
+            self.add_gate(&GateDescriptor::new());
+            self.get_entry_mut(i).unwrap().set_valid();
+            i = self.len()
+        }
+    }
+
+    /// Populates the [Table] with given [`GateDescriptor`] to reach 256 entries (required)
```
Comment (sleiderr) [inline src/bios/flib/src/idt/idt.rs:79]:
```suggestion
    /// Populates the `Table` with given [`GateDescriptor`] to reach 256 entries (required)
```

File: src/bios/flib/src/idt/idt.rs [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+    pub fn populate(&mut self, default : GateDescriptor) {
+        let mut i = self.len();
+        while i < 256 {
+            self.add_gate(&default);
+            i = self.len()
+        }
+    }
+
+
+
```
Comment (sleiderr) [inline src/bios/flib/src/idt/idt.rs:89]:
```suggestion
```

File: src/bios/flib/src/idt/idt.rs [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+/// |             index << 3                 | TI |  RPL  |
+///  -----------------------------------------------------
+pub struct SegmentSelector {
+    rpl: B2,
+    ti: B1,
+    index: B13,
+}
+
+impl SegmentSelector {
+    /// Returns the [SegmentSelector] configured to use loaded GDT
```
Comment (sleiderr) [inline src/bios/flib/src/idt/idt.rs:119]:
```suggestion
    /// Returns the `SegmentSelector` configured to use loaded GDT.
```

File: src/bios/flib/src/idt/idt.rs [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+    pub fn with_gdt(mut self) -> Self {
+        self.set_ti(0);
+        self
+    }
+
+    /// Returns the [SegmentSelector] configured with the given privileges
```
Comment (sleiderr) [inline src/bios/flib/src/idt/idt.rs:125]:
```suggestion
    /// Returns the `SegmentSelector` configured with the given privileges.
```

File: src/bios/flib/src/idt/idt.rs [hunk 16]
```diff
@@ -0,0 +1,229 @@
...
+    ti: B1,
+    index: B13,
+}
+
+impl SegmentSelector {
+    /// Returns the [SegmentSelector] configured to use loaded GDT
+    pub fn with_gdt(mut self) -> Self {
+        self.set_ti(0);
+        self
+    }
```
Comment (sleiderr) [inline src/bios/flib/src/idt/idt.rs:123]: pourquoi tu fais fn(mut self) -> Self et pas fn(&mut self) -> () ?

**Review by sleiderr — CHANGES_REQUESTED** [review]

File: src/bios/f-init32/src/interrupts/handlers.rs [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+/// It provides several utilities to define interrupts.
+/// To define a simple interrupt, precede your fn definiton with the proc
+/// macro #\[interrupt]
+/// # Example
+/// ```
+/// #[interrupt]
+/// pub fn _int0x00() {
+///     do_something()
+/// }
+/// ```
```
Comment (sleiderr) [inline src/bios/f-init32/src/interrupts/handlers.rs:20]:
```suggestion
///
/// # Examples:
///
/// ```
/// #[interrupt]
/// pub fn _int0x00() {
///     do_something()
/// }
/// ```
///
```

File: src/bios/f-init32/src/interrupts/handlers.rs [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+/// The naming convention implies that interrupt are named as follows :
```
Comment (sleiderr) [inline src/bios/f-init32/src/interrupts/handlers.rs:21]:
```suggestion
/// The naming convention implies that interrupt are named as follows :
///
```

File: src/bios/f-init32/src/interrupts/handlers.rs [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+/// ```
+/// let int_code : usize
+/// ```
+/// usable in your default template to perform generic handle actions at runtime.
+///```
+///#[interrupt_default]
+///pub fn _int_default(){
+///     print!("{}", int_code);
+///}
+///```
```
Comment (sleiderr) [inline src/bios/f-init32/src/interrupts/handlers.rs:38]:
```suggestion
/// This one will provide a generic variable :
/// ```
/// let int_code : usize
/// ```
///
/// usable in your default template to perform generic handle actions at runtime.
///
/// ```
/// #[interrupt_default]
/// pub fn _int_default(){
///      print!("{}", int_code);
/// }
/// ```
///
```

File: src/bios/f-init32/src/interrupts/handlers.rs [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+/// e.g. This example shows how you would save the idt at offset 0x8
+///```
+///#[interrupt_descriptor_table(0x8)]
+///mod interrupts;
+///```
+/// Loading of the table will then be achieved by calling the following function
+/// from your main function.
+/// ```
+/// generate_idt()
+/// ```
```
Comment (sleiderr) [inline src/bios/f-init32/src/interrupts/handlers.rs:54]:
```suggestion
/// e.g. This example shows how you would save the idt at offset 0x8
///
/// ```
/// #[interrupt_descriptor_table(0x8)]
/// mod interrupts;
/// ```
///
/// Loading of the table will then be achieved by calling the following function
/// from your main function.
///
/// ```
/// generate_idt()
/// ```
```

File: src/bios/f-init32/src/interrupts/handlers.rs [hunk 21]
```diff
@@ -0,0 +1,76 @@
+use alloc::boxed::Box;
+use alloc::format;
+use core::arch::asm;
+use core::mem::transmute;
+use f_macros::{interrupt, interrupt_default};
+use flib::debug::debug::print_str;
```
Comment (sleiderr) [inline src/bios/f-init32/src/interrupts/handlers.rs:6]:
```suggestion
```

File: src/bios/f-init32/src/interrupts/handlers.rs [hunk 22]
```diff
@@ -0,0 +1,76 @@
...
+    }
+
+    // scheduler_ref!(scheduler);
+    // let action = ScheduledAction::new(0x00);
+    // scheduler.schedule(action);
+}
+
+#[interrupt_default]
+pub fn _int_default() {
+    print_str(format!("{}", int_code).as_str(), 0);
```
Comment (sleiderr) [inline src/bios/f-init32/src/interrupts/handlers.rs:75]:
```suggestion
    println!("{}", int_code);
```

File: src/bios/f-init32/src/interrupts/handlers.rs [hunk 23]
```diff
@@ -0,0 +1,76 @@
+use alloc::boxed::Box;
+use alloc::format;
+use core::arch::asm;
+use core::mem::transmute;
+use f_macros::{interrupt, interrupt_default};
+use flib::debug::debug::print_str;
+use flib::int::scheduler::{IntScheduler, ScheduledAction};
+use flib::{print, scheduler_ref};
```
Comment (sleiderr) [inline src/bios/f-init32/src/interrupts/handlers.rs:8]:
```suggestion
use flib::{println, scheduler_ref};
```

File: src/bios/f-macros/src/lib.rs [hunk 24]
```diff
@@ -0,0 +1,227 @@
...
+use syn::{Ident, Item, ItemFn, parse_file, parse_macro_input, Stmt};
+use syn::__private::str;
+use syn::spanned::Spanned;
+
+/// This procedural macro will generate a function that will build the IDT from
+/// the module where all interrupts are defined.
+/// This function will then have to be called from the main function.
+/// ```
+/// generate_idt();
+/// ```
```
Comment (sleiderr) [inline src/bios/f-macros/src/lib.rs:20]:
```suggestion
/// This function will then have to be called from the main function.
///
/// ```
/// generate_idt();
/// ```
```

File: src/bios/flib/src/lib.rs [hunk 25]
```diff
@@ -10,9 +10,22 @@ pub mod mem;
...
 
 #[cfg(feature = "alloc")]
 extern crate alloc;
 
 extern crate rlibc;
+
+#[cfg(feature = "alloc")]
+extern crate alloc;
+
+
```
Comment (sleiderr) [inline src/bios/flib/src/lib.rs:31]:
```suggestion
```

File: src/bios/flib/src/idt/idt.rs [hunk 26]
```diff
@@ -0,0 +1,229 @@
...
+use modular_bitfield;
+use modular_bitfield::bitfield;
+use modular_bitfield::BitfieldSpecifier;
+use modular_bitfield::prelude::{B1, B11, B13, B2, B4, B8};
+
+use crate::debug;
+use crate::debug::debug::print_str;
+use crate::gdt::gdt::SegmentDescriptor;
+
+use numtoa::NumToA;
```
Comment (sleiderr) [inline src/bios/flib/src/idt/idt.rs:14]: Do you still use `numtoa`? didn't see it used anywhere, and we shouldn't need that anymore (at least in protected mode) with the formatting macros.

Comment (raphfaur) [reply src/bios/flib/src/idt/idt.rs:14]: You're right, I was using it to debug on protected mode before vesa got merged
```suggestion
```

File: src/bios/flib/src/idt/idt.rs (hunk 16 above)
Comment (raphfaur) [reply src/bios/flib/src/idt/idt.rs:123]: It is useful for chaining when creating SegmentSelectors

File: src/bios/f-init32/Cargo.toml [hunk 27]
```diff
@@ -7,6 +7,8 @@ edition = "2021"
 
 [dependencies]
 flib = { path = "../flib/", features = ["alloc"] }
+f-macros = { path = "../f-macros/" }
+numtoa = "0.2"
```
Comment (sleiderr) [inline src/bios/f-init32/Cargo.toml:11]:
```suggestion
```

File: src/bios/flib/src/idt/idt.rs (hunk 16 above)
Comment (raphfaur) [reply src/bios/flib/src/idt/idt.rs:119]: overleve style

File: src/bios/flib/src/io/utils.rs
Comment (sleiderr) [inline src/bios/flib/src/io/utils.rs:1]: c'était déjà implémenté

File: src/bios/flib/src/idt/idt.rs [hunk 28]
```diff
@@ -0,0 +1,221 @@
...
+    }
+
+    /// Stores the [IDTDescriptor] to a given location in memory
+    pub fn store(&self, offset: usize) {
+        let ptr = offset as *mut IDTDescriptor;
+        unsafe { write_volatile(ptr, self.clone()) }
+    }
+}
+
+/// Loads [IDTDescriptor] in the CPU IDTR using `lidt` instruction
```
Comment (sleiderr) [inline src/bios/flib/src/idt/idt.rs:211]:
```suggestion
/// Loads [`IDTDescriptor`] in the CPU IDTR using `lidt` instruction
```

File: src/bios/flib/src/idt/idt.rs [hunk 29]
```diff
@@ -0,0 +1,221 @@
...
+            offset: 0x00,
+        }
+    }
+
+    /// Set offset of the [`IDTDescriptor`]
+    pub fn set_offset(&mut self, offset: u32) {
+        self.offset = offset
+    }
+
+    /// Stores the [IDTDescriptor] to a given location in memory
```
Comment (sleiderr) [inline src/bios/flib/src/idt/idt.rs:204]:
```suggestion
    /// Stores the [`IDTDescriptor`] to a given location in memory
```

File: src/bios/flib/src/idt/idt.rs [hunk 30]
```diff
@@ -0,0 +1,221 @@
...
+            (*bytes.get(2).unwrap() as u16) + ((*bytes.get(3).unwrap() as u16) << 8),
+        );
+    }
+
+    /// Set gate type
+    pub fn set_type(&mut self, gt: GateType) {
+        self.set_gate_type(gt as u8);
+    }
+
+    /// Set p bit to 1. One must always call [set_valid()] on a [GateDescriptor]
```
Comment (sleiderr) [inline src/bios/flib/src/idt/idt.rs:165]:
```suggestion
    /// Set p bit to 1. One must always call `set_valid()` on a [`GateDescriptor`]
```

File: src/bios/flib/src/idt/idt.rs [hunk 31]
```diff
@@ -0,0 +1,221 @@
...
+    /// Adds a [GateDescriptor] to the [Table]
+    pub fn add_gate(&mut self, gate: &GateDescriptor) {
+        self.entries.push(gate.clone());
+    }
+}
+
+#[repr(C, packed)]
+#[bitfield]
+#[derive(BitfieldSpecifier)]
+/// The [SegmentSelector] describes the segment where the Routine is stored.
```
Comment (sleiderr) [inline src/bios/flib/src/idt/idt.rs:98]:
```suggestion
/// The `SegmentSelector` describes the segment where the Routine is stored.
```

**Review by sleiderr — APPROVED** [review]

**Pull request merged**